### ISSUE TITLE
`signFields` method generates signatures not safely verifiable on contract side due to JSON parsing issue

### ISSUE THREAD

**Issue by ffilipovicc98** [aurowallet/auro-wallet-browser-extension#22]
Hello 👋 

It’s great to see that `signFields` has been added as a new method in the injected Mina provider. I’ve been trying to use it and I’ve found that while simple fields can be verified, some more complex fields are not possible to verify safely on the contract side. In the example below, `verifySignature1` is safe while `verifySignature2` is unsafe.
```js
const simpleField = snarkyjs.Field(12);

const complexField = snarkyjs.Poseidon.hash([]);
```

I’ve created a few tests and a test project that demonstrate how I use `signFields` and how I’ve tried to validate signatures generated by `signFields`. To run the tests, follow these steps:

1. Make sure you have the Auro wallet installed.
2. Open this [url](https://sign-fields-issue.web.app/).
3. Open the console tab in developer tools.
4. Paste this code block:

```js
await (async () => {
  // possible on the contract
  const verifySignature1 = (signature, field, publicKey) => {
    try {
      signature
        .verify(
          snarkyjs.PublicKey.fromBase58(publicKey),
          field.toFields().map(snarkyjs.Field),
        )
        .assertTrue();

      console.log("Valid signature");
    } catch (err) {
      console.log(err.message);
      console.log("Invalid signature");
    }
  };

  // impossible on the contract
  const verifySignature2 = (signature, field, publicKey) => {
    try {
      signature
        .verify(
          snarkyjs.PublicKey.fromBase58(publicKey),
          JSON.parse(`[${field.toFields()}]`)
          .map(BigInt)
          .map(snarkyjs.Field),
          // or
          // field.toFields().map(el => snarkyjs.Field(BigInt(+el.toString())))
        )
        .assertTrue();

      console.log("Valid signature");
    } catch (err) {
      console.log(err.message);
      console.log("Invalid signature");
    }
  };

  // setup
  const client = new Client({
    network: "testnet"
  });
  const privateKey = snarkyjs.PrivateKey.random();
  const publicKey = privateKey.toPublicKey().toBase58();
  const simpleField = snarkyjs.Field(12);
  const complexField = snarkyjs.Poseidon.hash([]);

  // tests
  const test1 = async () => {
    const signature = snarkyjs.Signature.fromBase58(
      client.signFields(
        simpleField.toFields().map(BigInt),
        privateKey.toBase58(),
      ).signature,
    );

    verifySignature1(signature, simpleField, publicKey);
  };

  const test2 = async () => {
    const signature = snarkyjs.Signature.fromBase58(
      client.signFields(
        complexField.toFields().map(BigInt),
        privateKey.toBase58(),
      ).signature,
    );

    verifySignature1(signature, complexField, publicKey);
  };

  const test3 = async () => {
    const publicKey = (await window.mina.requestAccounts())[0];

    const signature = snarkyjs.Signature.fromBase58(
      (
        await window.mina.signFields({
          message: `[${simpleField.toFields()}]`,
        })
      ).signature,
    );

    verifySignature1(signature, simpleField, publicKey);
  };

  const test4 = async () => {
    const publicKey = (await window.mina.requestAccounts())[0];

    const signature = snarkyjs.Signature.fromBase58(
      (
        await window.mina.signFields({
          message: `[${complexField.toFields()}]`,
        })
      ).signature,
    );

    verifySignature1(signature, complexField, publicKey);
  };

  const test5 = async () => {
    const publicKey = (await window.mina.requestAccounts())[0];

    const signature = snarkyjs.Signature.fromBase58(
      (
        await window.mina.signFields({
          message: `[${simpleField.toFields()}]`,
        })
      ).signature,
    );

    verifySignature2(signature, simpleField, publicKey);
  };

  const test6 = async () => {
    const publicKey = (await window.mina.requestAccounts())[0];

    const signature = snarkyjs.Signature.fromBase58(
      (
        await window.mina.signFields({
          message: `[${complexField.toFields()}]`,
        })
      ).signature,
    );

    verifySignature2(signature, complexField, publicKey);
  };

  (async () => {
    console.log("+---- test1 ----+");
    await test1();

    console.log("\n+---- test2 ----+");
    await test2();

    console.log("\n+---- test3 ----+");
    await test3();

    console.log("\n+---- test4 ----+");
    await test4();

    console.log("\n+---- test5 ----+");
    await test5();

    console.log("\n+---- test6 ----+");
    await test6();
  })();
})();
```

All tests are successful except for `test4`. What these tests show is that I can successfully sign and verify both simple and complex fields using the Mina signer. However, the Auro wallet uses `JSON.parse` in its `signFields` implementation, which causes verification to fail on some complex fields.

I ended up with this code
```js
snarkyjs.Signature.fromBase58(
  (
    await window.mina.signFields({
      message: `[${simpleField.toFields()}]`,
    })
  ).signature,
);
```

by looking at the implementation code of [`signFields`](https://github.com/aurowallet/auro-wallet-browser-extension/blob/1372580f51e5dc1f75001aa1625b7c71d402dcd1/src/background/lib/index.js#L165)
```js
export async function signFieldsMessage(privateKey,params){
    let signResult
    try {
        const fields = JSON.parse(params.message)
        if(!Array.isArray(fields)){
           return { error: { message: "Must use array" } }
        }
        const nextFields = fields.map((field)=>{
            if(typeof field === "number"){
                return BigInt(field)
            }
            return field
        })
        let signClient = await getSignClient()
        signResult = signClient.signFields(nextFields, privateKey)
        signResult.data = fields
    } catch (err) {
        let errorMessage = getRealErrorMsg(err)||i18n.t("buildFailed")
        signResult = { error: { message: errorMessage } }
    }
    return signResult
}
```

To summarize, while `signFields` works for simple fields, it doesn’t seem to work reliably for complex fields. Could you take a look at this and see if there’s a way to make it work for all field types? Thanks!

OS: macOS Monterey v12.1
Browser: Chrome 112.0.5615.137
snarkyjs: 0.9.8
mina-signer: 2.0.3
auro-wallet-chrome-extension: 2.1.10

**Comment by lvshaoping007** [comment]
We fixed this by changing the parameter type and how it is passed. You can try the latest package, [here](https://github.com/aurowallet/auro-wallet-browser-extension/blob/master/src/background/lib/index.js#L169) is the modified core

**Comment by ffilipovicc98** [comment]
It seems that this issue hasn't been resolved adequately, as it lacks a clear explanation on how to utilize the new API. I'm having difficulty understanding the process of transforming `snarkyjs.Field` into an array of numbers. Could you please assist me by modifying the `message` parameter passed to `signFields` in my `test3` and `test4` so that they can be successfully verified using `verifySignature1`? It would be greatly appreciated. Thank you.

If you're facing difficulties running the tests, I suggest arranging a call so that we can collaborate and resolve the issue together.

**Comment by ffilipovicc98** [comment]
Maybe `signFields` could be implemented:

```js
export async function signFieldsMessage(privateKey,params){
    let signResult
    try {
        const nextFields = snarkyjs.Field.fromJSON(params.message).toFields().map(BigInt)
        let signClient = await getSignClient()
        signResult = signClient.signFields(nextFields, privateKey)
        signResult.data = nextFields
    } catch (err) {
        let errorMessage = getRealErrorMsg(err)||i18n.t("buildFailed")
        signResult = { error: { message: errorMessage } }
    }
    return signResult
}
```

And used: 

```js
await window.mina.signFields({
    message: simpleField.toJSON(),
})
```

**Comment by wjdfx** [comment]
@ffilipovicc98 How can I get in touch with you? My discord handle is **niu8** can you add me?

**Comment by wjdfx** [comment]
@ffilipovicc98 How can I get in touch with you? My discord handle is **niu8** can you add me?

**Comment by ffilipovicc98** [comment]
I have recently discovered the correct way to utilize it thanks to the assistance of @mitschabaude. Here is the full working example, demonstrating how to generate signatures on the frontend and how to verify them on the contracts.

```js
await (async () => {
  // possible on the contract
  const verifySignature1 = (signature, field, publicKey) => {
    try {
      signature
        .verify(
          snarkyjs.PublicKey.fromBase58(publicKey),
          field.toFields().map(snarkyjs.Field),
        )
        .assertTrue();

      console.log("Valid signature");
    } catch (err) {
      console.log(err.message);
      console.log("Invalid signature");
    }
  };

  const privateKey = snarkyjs.PrivateKey.random();
  const publicKey = privateKey.toPublicKey().toBase58();
  const simpleField = snarkyjs.Field(12);
  const complexField = snarkyjs.Poseidon.hash([]);

  const test3 = async () => {
    const publicKey = (await window.mina.requestAccounts())[0];
      
    const signature = snarkyjs.Signature.fromBase58(
      (
        await window.mina.signFields({
          message: [simpleField.toString()],
        })
      ).signature,
    );

    verifySignature1(signature, simpleField, publicKey);
  };

  const test4 = async () => {
    const publicKey = (await window.mina.requestAccounts())[0];

    const signature = snarkyjs.Signature.fromBase58(
      (
        await window.mina.signFields({
          message: [complexField.toString()],
        })
      ).signature,
    );

    verifySignature1(signature, complexField, publicKey);
  };

  (async () => {
    console.log("\n+---- test3 ----+");
    await test3();

    console.log("\n+---- test4 ----+");
    await test4();
  })();
    
})();
```
